mermithid/processors/misc/BinnedDataFitter.py

```diff
@@ -0,0 +1,169 @@
+'''
+Author: C. Claessens
+Date:6/12/2020
+Description: Configure with arbitrary histogram and pdf.
+Minimizes poisson loglikelihood using iMinuit.
+
+'''
+
+from __future__ import absolute_import
+
+import numpy as np
+import scipy
+import sys
+import json
+from iminuit import Minuit
+from morpho.utilities import morphologging, reader
+from morpho.processors import BaseProcessor
+
+logger = morphologging.getLogger(__name__)
+
+
+
+__all__ = []
+__all__.append(__name__)
+
+class BinnedDataFitter(BaseProcessor):
+    '''
+    Processor that
+    Args:
+        variables
+        parameter_names
+        initial_values
+        limits
+        fixed
+        bins
+        print_level
+        constrained_parameter_indices
+        constrained_parameter_means
+        constrained_parameter_widths
+
+    Inputs:
+        data:
+    Output:
+        result: dictionary containing
+    '''
+    def InternalConfigure(self, params):
+        '''
+        Configure
+        '''
+
+        # Read other parameters
+        self.namedata = reader.read_param(params, 'variables', "required")
+        self.parameter_names = reader.read_param(params, 'parameter_names', ['A', 'mu', 'sigma'])
+        self.initial_values = reader.read_param(params, 'initial_values', [1, 0, 1])
+        self.limits = reader.read_param(params, 'limits', [[None, None], [None, None], [None, None]])
+        self.fixes = reader.read_param(params, 'fixed', [False, False, False])
+        self.bins = reader.read_param(params, 'bins', np.linspace(-2, 2, 100))
+        self.print_level = reader.read_param(params, 'print_level', 1)
+
+        self.constrained_parameters = reader.read_param(params, 'constrained_parameter_indices', [])
+        self.constrained_means = reader.read_param(params, 'constrained_parameter_means', [])
+        self.constrained_widths = reader.read_param(params, 'constrained_parameter_widths', [])
+
+        # derived configurations
+        self.bin_centers = self.bins[0:-1]+0.5*(self.bins[1]-self.bins[0])
+        self.parameter_errors = [max([0.1, 0.1*p]) for p in self.initial_values]
+
+        return True
+
+    def InternalRun(self):
+        logger.info('namedata: {}'.format(self.namedata))
+
+        self.hist, _ = np.histogram(self.data[self.namedata], self.bins)
+
+        result_array, error_array = self.fit()
+
+        # save results
+        self.results = {}
+        self.results['param_values'] = result_array
+        self.results['param_errors'] = error_array
+        for i, k in enumerate(self.parameter_names):
+            self.results[k] = {'value': result_array[i], 'error': error_array[i]}
+
+        return True
+
+
+
+    def PDF(self, x, A, mu, sigma):
+        """
+        Overwrite by whatever PDF
+        """
+        f = A*(1/(sigma*np.sqrt(2*np.pi)))*np.exp(-(((x-mu)/sigma)**2.)/2.)
```

Comment:
Here it is the case; but is there a way to check the PDF is normalized to 1?
If not and one uses unnormalized pdf (or at least have a non-constant norm), fits can go very wrong...

Reply:
don't know how I should do that. this method is supposed to be overwritten by the user. and I wanted to avoid making rules for the user like: the first parameter has to be the amplitude and the rest has to be normalized to 1.

Reply:
I understand your point! Maybe then a sentence about this need in the documentation would be useful...